### PR TITLE
Implemented Ishftc

### LINKED ISSUE #3083: implement intrinsic `ishftc`
```fortran
program main
    integer(8) :: st(4) = [1, 2, 3, 4]
    print*, ishftc(st(2) * 5, 2)
end
```
```console
$ gfortran b.f90 && ./a.out
                   40
$ lfortran b.f90
semantic error: Function 'ishftc' not found or not implemented yet (if it is intrinsic)
 --> b.f90:3:13
  |
3 |     print*, ishftc(st(2) * 5, 2)
  |             ^^^^^^^^^^^^^^^^^^^^ 


Note: Please report unclear or confusing messages as bugs at
https://github.com/lfortran/lfortran/issues.
```
Ref: https://github.com/lfortran/lfortran/issues/3052#issuecomment-1886572426

### PR DESCRIPTION
**Pull request by parth121101**
fix #3083

### CODE REVIEW

File: src/libasr/pass/intrinsic_function_registry.h [hunk 1]
```diff
@@ -2306,6 +2308,103 @@ namespace Idint {
...
+                    iPow(
+                        two,
+                        i_tSub(kind_expr, args[1], arg_types[0]),
+                        arg_types[0]
+                    ),
+                    arg_types[0]
+                ),
+                loc
+            )
+        ));
```
Comment (HarshitaKalani) [inline src/libasr/pass/intrinsic_function_registry.h:2397]: I think you can write this in compressed form only.

File: src/libasr/pass/intrinsic_function_registry.h [hunk 2]
```diff
@@ -2306,6 +2308,103 @@ namespace Idint {
...
+        ASR::expr_t *m_value = nullptr;
+        if (all_args_evaluated(args)) {
+            Vec<ASR::expr_t*> arg_values; arg_values.reserve(al, 2);
+            arg_values.push_back(al, expr_value(args[0]));
+            arg_values.push_back(al, expr_value(args[1]));
+            int64_t shift = ASR::down_cast<ASR::IntegerConstant_t>(arg_values[1])->m_n;
+            int kind = ASRUtils::extract_kind_from_ttype_t(ASR::down_cast<ASR::IntegerConstant_t>(arg_values[0])->m_type);
+            int bits = 8*kind;
+            if (bits < shift) {
+                err("The absolute value of SHIFT at (1) must be less than or equal to BIT_SIZE('I')",
```
Comment (HarshitaKalani) [inline src/libasr/pass/intrinsic_function_registry.h:2360]:
```suggestion
                err("The absolute value of SHIFT argument must be less than or equal to BIT_SIZE('I')",
```

File: src/libasr/pass/intrinsic_function_registry.h [hunk 3]
```diff
@@ -2306,6 +2308,103 @@ namespace Idint {
...
+
+    static ASR::expr_t *eval_Ishftc(Allocator &al, const Location &loc,
+            ASR::ttype_t* t1, Vec<ASR::expr_t*> &args) {
+        int64_t val = ASR::down_cast<ASR::IntegerConstant_t>(args[0])->m_n;
+        int64_t shift = ASR::down_cast<ASR::IntegerConstant_t>(args[1])->m_n;
+        int kind = 8*ASRUtils::extract_kind_from_ttype_t(ASR::down_cast<ASR::IntegerConstant_t>(args[0])->m_type);
+        int result;
+        if (shift < 0) {
+            shift = -shift;  
+            result = (val >>shift) | (val << (kind - shift));
```
Comment (HarshitaKalani) [inline src/libasr/pass/intrinsic_function_registry.h:2332]:
```suggestion
            result = (val >> shift) | (val << (kind - shift));
```

File: integration_tests/intrinsics_131.f90 [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+    integer :: a, b
+    a = 10
+    b = 2
+    print *, ishftc(a, b)
+    if (ishftc(a, b) /= 40) error stop
+
+    print *, ishftc(10, 2)
+    if (ishftc(10, 2) /= 40) error stop
+
+end program
```
Comment (HarshitaKalani) [inline integration_tests/intrinsics_131.f90:11]: Can you please add tests checking kind = 8 too?

File: src/libasr/pass/intrinsic_function_registry.h [hunk 5]
```diff
@@ -2306,6 +2308,85 @@ namespace Idint {
...
+        ASRUtils::require_impl((is_integer(*type1) && is_integer(*type2)),
+            "Arguments to `Ishftc` must be of integer type",
+            x.base.base.loc, diagnostics);
+    }
+
+    static ASR::expr_t *eval_Ishftc(Allocator &al, const Location &loc,
+            ASR::ttype_t* t1, Vec<ASR::expr_t*> &args) {
+        int64_t val = ASR::down_cast<ASR::IntegerConstant_t>(args[0])->m_n;
+        int64_t shift = ASR::down_cast<ASR::IntegerConstant_t>(args[1])->m_n;
+        int kind = 8*ASRUtils::extract_kind_from_ttype_t(ASR::down_cast<ASR::IntegerConstant_t>(args[0])->m_type);
```
Comment (Shaikh-Ubaid) [inline src/libasr/pass/intrinsic_function_registry.h:2328]: I think this should be `bitsize` (or `no_of_bits` or something similar). For example we can have:
```suggestion
        int kind = ASRUtils::extract_kind_from_ttype_t(ASR::down_cast<ASR::IntegerConstant_t>(args[0])->m_type);
        int bitsize = 8 * kind;
```

and then reuse `bitsize` whereever needed.

File: src/libasr/pass/intrinsic_function_registry.h [hunk 6]
```diff
@@ -2306,6 +2308,85 @@ namespace Idint {
...
+            arg_values.push_back(al, expr_value(args[0]));
+            arg_values.push_back(al, expr_value(args[1]));
+            int64_t shift = ASR::down_cast<ASR::IntegerConstant_t>(arg_values[1])->m_n;
+            int kind = ASRUtils::extract_kind_from_ttype_t(ASR::down_cast<ASR::IntegerConstant_t>(arg_values[0])->m_type);
+            int bits = 8*kind;
+            if (bits < shift) {
+                err("The absolute value of SHIFT argument must be less than or equal to BIT_SIZE('I')",
+                args[0]->base.loc);
+            }
+            m_value = eval_Ishftc(al, loc, expr_type(args[0]), arg_values);
```
Comment (Shaikh-Ubaid) [inline src/libasr/pass/intrinsic_function_registry.h:2363]: I think the above checking of `bits < shift` should be done inside the `eval_Ishftc()`.

File: src/libasr/pass/intrinsic_function_registry.h [hunk 7]
```diff
@@ -2306,6 +2308,85 @@ namespace Idint {
...
+    }
+
+    static inline ASR::expr_t* instantiate_Ishftc(Allocator &al, const Location &loc,
+            SymbolTable *scope, Vec<ASR::ttype_t*>& arg_types, ASR::ttype_t *return_type,
+            Vec<ASR::call_arg_t>& new_args, int64_t /*overload_id*/) {
+        declare_basic_variables("_lcompilers_Ishftc_" + type_to_str_python(arg_types[0]));
+        fill_func_arg("x", arg_types[0]);
+        fill_func_arg("y", arg_types[1]);
+        auto result = declare(fn_name, return_type, ReturnVar);
+        int kind = 8*ASRUtils::extract_kind_from_ttype_t(arg_types[0]);
```
Comment (Shaikh-Ubaid) [inline src/libasr/pass/intrinsic_function_registry.h:2377]:
```suggestion
        int kind = ASRUtils::extract_kind_from_ttype_t(arg_types[0]);
        int no_of_bits = 8 * kind;
```

and then reuse `no_of_bits` (or `bitsize` or whichever name is more appropriate).

File: integration_tests/intrinsics_131.f90 [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+    print *, ishftc(c, d)
+    if (ishftc(c, d) /= 40_8) error stop
+
+    print *, ishftc(10, 2)
+    if (ishftc(10, 2) /= 40) error stop
+
+    print *, ishftc(10_8, 2_8)
+    if (ishftc(10_8, 2_8) /= 40_8) error stop
+
+end program
```
Comment (Shaikh-Ubaid) [inline integration_tests/intrinsics_131.f90:21]: Can we also add tests for `0` and negative shifts as well? I think this would make it a more robust test.

**Review by Shaikh-Ubaid — COMMENTED** [review]
Great Work! Thanks! I shared some comments.

File: src/libasr/pass/intrinsic_function_registry.h (hunk 6 above)
Comment (parth121101) [reply src/libasr/pass/intrinsic_function_registry.h:2363]: There is no way to throw error in eval_Ishftc.

File: src/libasr/pass/intrinsic_function_registry.h [hunk 9]
```diff
@@ -2306,6 +2308,85 @@ namespace Idint {
...
+            args.p, args.n, 0, ASRUtils::expr_type(args[0]), m_value);
+    }
+
+    static inline ASR::expr_t* instantiate_Ishftc(Allocator &al, const Location &loc,
+            SymbolTable *scope, Vec<ASR::ttype_t*>& arg_types, ASR::ttype_t *return_type,
+            Vec<ASR::call_arg_t>& new_args, int64_t /*overload_id*/) {
+        declare_basic_variables("_lcompilers_Ishftc_" + type_to_str_python(arg_types[0]));
+        fill_func_arg("x", arg_types[0]);
+        fill_func_arg("y", arg_types[1]);
+        auto result = declare(fn_name, return_type, ReturnVar);
```
Comment (Shaikh-Ubaid) [inline src/libasr/pass/intrinsic_function_registry.h:2376]: For clarity you can also add a comment here about the steps the function would take to compute the results. For example, see https://github.com/lfortran/lfortran/blob/1534e8601340d743e9a7df038a5d5a3a4a22a664/src/libasr/pass/intrinsic_function_registry.h#L1812-L1815

**Comment by certik** [comment]
@parth121101 once you address the feedback, please click on "Ready for review".

File: integration_tests/intrinsics_131.f90 (hunk 8 above)
Comment (parth121101) [reply integration_tests/intrinsics_131.f90:21]:
```
$ cat b.f90
program main
    print *, ishftc(10, -2)

end program

$ gfortran b.f90 && ./a.out
 -2147483646
```

gfortran is not supporting negative shift I guess. 
But lfortran is giving true results.

Comment (Shaikh-Ubaid) [reply integration_tests/intrinsics_131.f90:21]: As per [docs](https://gcc.gnu.org/onlinedocs/gfortran/ISHFTC.html), `ishftc` supports negative shifts. It also works for me locally.

```console
% cat examples/expr2.f90 
program main
    print *, ishftc(10, -1)

end program
% gfortran examples/expr2.f90 
% ./a.out 
           5
```
We need to match the output of `gfortran`. If not, I guess there could be bug somewhere.

Comment (certik) [reply integration_tests/intrinsics_131.f90:21]: GFortran is most likely correct. We need to implement these intrinsics in the way other compilers do it, as well as the standard prescribes.

File: src/libasr/pass/intrinsic_function_registry.h [hunk 10]
```diff
@@ -2306,6 +2308,77 @@ namespace Idint {
...
+        /* 
+         * r = ishftc(x, y) 
+         * r = (x << y) | (x >> (bits_size - y))
```
Comment (Shaikh-Ubaid) [inline src/libasr/pass/intrinsic_function_registry.h:2366]: Please mention here the steps the function is taking to compute the result. Below on line `2372`, it seems to use `i_tMul`, `i_tDiv`, `iPow` and operations among them to compute the result. I would mention that exact equation using symbols `*`, `/`, `**`, etc. in the comment here.

File: src/libasr/pass/intrinsic_function_registry.h [hunk 11]
```diff
@@ -2306,6 +2308,77 @@ namespace Idint {
...
+            ASR::ttype_t* t1, Vec<ASR::expr_t*> &args) {
+        int64_t val = ASR::down_cast<ASR::IntegerConstant_t>(args[0])->m_n;
+        int64_t shift = ASR::down_cast<ASR::IntegerConstant_t>(args[1])->m_n;
+        int kind = ASRUtils::extract_kind_from_ttype_t(ASR::down_cast<ASR::IntegerConstant_t>(args[0])->m_type);
+        int bits_size = 8*kind;
+        if (bits_size < shift) {
+            throw SemanticError("The absolute value of SHIFT argument must be less than or equal to BIT_SIZE('I')",
+            args[0]->base.loc); 
+        }
+        int result;
```
Comment (Shaikh-Ubaid) [inline src/libasr/pass/intrinsic_function_registry.h:2323]: I think this result would overflow for kind `8` integers. It probably needs to be of type `int64_t`.

File: integration_tests/intrinsics_131.f90 [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+    if (ishftc(c, d) /= 40_8) error stop
+
+    print *, ishftc(10, 2)
+    if (ishftc(10, 2) /= 40) error stop
+
+    print *, ishftc(10_8, 2_8)
+    if (ishftc(10_8, 2_8) /= 40_8) error stop
+
+    print *, ishftc(10, -1)
+    if (ishftc(10, -1) /= 5) error stop
```
Comment (Shaikh-Ubaid) [inline integration_tests/intrinsics_131.f90:22]: > ```
> $ cat b.f90
> program main
>     print *, ishftc(10, -2)
> 
> end program
> 
> $ gfortran b.f90 && ./a.out
>  -2147483646
> ```
> 
> gfortran is not supporting negative shift I guess. But lfortran is giving true results.

Can we add the test for `ishftc(10, -2)` (kind 4 as well as kind 8) with compile-time and run-time values? I think checking `ishftc(10, -2)` would ensure if the circulatory implementation of `ishftc()` is working correctly.

**Review by Shaikh-Ubaid — COMMENTED** [review]
Thanks! I shared few minor comments. After resolving that this should be good to merge.

File: integration_tests/intrinsics_131.f90 (hunk 12 above)
Comment (parth121101) [reply integration_tests/intrinsics_131.f90:22]: I will apply code review tomorrow morning as soon as possible.

**Comment by parth121101** [comment]
I have applied the code review. Can we implement negative shift cases in another PR as of now?

File: src/libasr/pass/intrinsic_function_registry.h [hunk 13]
```diff
@@ -2306,6 +2308,77 @@ namespace Idint {
...
+    static inline ASR::expr_t* instantiate_Ishftc(Allocator &al, const Location &loc,
+            SymbolTable *scope, Vec<ASR::ttype_t*>& arg_types, ASR::ttype_t *return_type,
+            Vec<ASR::call_arg_t>& new_args, int64_t /*overload_id*/) {
+        declare_basic_variables("_lcompilers_Ishftc_" + type_to_str_python(arg_types[0]));
+        fill_func_arg("x", arg_types[0]);
+        fill_func_arg("y", arg_types[1]);
+        auto result = declare(fn_name, return_type, ReturnVar);
+        /* 
+         * r = ishftc(x, y) 
+         * r = (x*(2**y)) | (x/(2**(kind(x)-y)))
```
Comment (Shaikh-Ubaid) [inline src/libasr/pass/intrinsic_function_registry.h:2366]: I think this should be 

```suggestion
         * r = (x*(2**y)) | (x/(2**((8 * kind(x))-y)))
```
or
```suggestion
         * r = (x*(2**y)) | (x/(2**(bitsize(x)-y)))
```

File: src/libasr/pass/intrinsic_function_registry.h [hunk 14]
```diff
@@ -2306,6 +2308,77 @@ namespace Idint {
...
+        auto result = declare(fn_name, return_type, ReturnVar);
+        /* 
+         * r = ishftc(x, y) 
+         * r =  r = (x*(2**y)) | (x/(2**(bitsize(x)-y)))
+         */ 
+        int kind = ASRUtils::extract_kind_from_ttype_t(arg_types[0]);
+        int bits_size = 8*kind;
+        ASR::expr_t *two = i(2, arg_types[0]);
+        ASR::expr_t *kind_expr = i(bits_size, arg_types[0]);
+        body.push_back(al, b.Assignment(result,b.Or(i_tMul(args[0],iPow(two, args[1], arg_types[0]), arg_types[0]),i_tDiv(args[0],iPow(two,i_tSub(kind_expr, args[1], arg_types[0]),arg_types[0]),arg_types[0]),loc)));
```
Comment (certik) [inline src/libasr/pass/intrinsic_function_registry.h:2372]: Let's use ASR intrinsics for bit shifting: https://github.com/lfortran/lfortran/issues/3247

**Comment by certik** [comment]
I think it's time to figure out the correct behavior of these shift functions for both positive and negative arguments and get implement it once and correctly. Then move to the next function.

File: src/libasr/pass/intrinsic_function_registry.h [hunk 15]
```diff
@@ -2306,6 +2308,77 @@ namespace Idint {
...
+            throw SemanticError("The absolute value of SHIFT argument must be less than or equal to BIT_SIZE('I')",
+            args[0]->base.loc); 
+        }
+        int64_t result;
+        if (shift < 0) {
+            shift = -shift;  
+            result = (val >> shift) | (val << (bits_size - shift));
+        } else {
+            result = (val << shift) | (val >> (bits_size - shift)); 
+        }
```
Comment (certik) [inline src/libasr/pass/intrinsic_function_registry.h:2329]: Is this the correct implementation for all positive and negative arguments?

If yes: then let's implement this for runtime in instantiate_Ishftc as well and be done with this instrinsic.

If not: then let's not commit this code if it is incorrect, and rather let's figure out the correct implementation, and then implement it both for compile time and runtime, and add proper tests for both.

**Comment by Pranavchiku** [comment]
Left circular shift implementation in cpp: https://en.cppreference.com/w/cpp/numeric/rotl

File: src/libasr/pass/intrinsic_function_registry.h (hunk 15 above)
Comment (Pranavchiku) [reply src/libasr/pass/intrinsic_function_registry.h:2329]: I looked at this, and found it is correct for positive value and negative shift but, not for negative values. I tried following two's complement approach but it didn't work for me. The cpp `rotr` and `rotl` is for unsigned types. We'll have to find correct approach for signed numbers.

```cpp
        if ( val < 0 ) {
            // take twos complement
            val = ~val + 1;
            is_negative = true;
        }
        if (shift < 0) {
            shift = -shift;
            result = (val >> shift) | (val << (bits_size - shift));
        } else {
            result = (val << shift) | (val >> (bits_size - shift)); 
        }
        if (is_negative) {
            result = ~result + 1;
            std::cout<<"result "<<result<<std::endl;
        }
```

Comment (Shaikh-Ubaid) [reply src/libasr/pass/intrinsic_function_registry.h:2329]: The following approach works:

```diff
diff --git a/src/libasr/pass/intrinsic_function_registry.h b/src/libasr/pass/intrinsic_func
tion_registry.h
index dd26f7fbe..0dcab47d6 100644
--- a/src/libasr/pass/intrinsic_function_registry.h
+++ b/src/libasr/pass/intrinsic_function_registry.h
@@ -2310,22 +2310,34 @@ namespace Idint {
 
 namespace Ishftc {
 
+    static uint64_t cutoff_extra_bits(uint64_t num, uint32_t bits_size, uint32_t max_bits_
size) {
+        if (bits_size == max_bits_size) {
+            return num;
+        }
+        return (num & ((1lu << bits_size) - 1lu));
+    }
+
     static ASR::expr_t *eval_Ishftc(Allocator &al, const Location &loc,
             ASR::ttype_t* t1, Vec<ASR::expr_t*> &args) {
-        int64_t val = ASR::down_cast<ASR::IntegerConstant_t>(args[0])->m_n;
-        int64_t shift = ASR::down_cast<ASR::IntegerConstant_t>(args[1])->m_n;
+        uint64_t val = (uint64_t)ASR::down_cast<ASR::IntegerConstant_t>(args[0])->m_n;
+        int64_t shift_signed = ASR::down_cast<ASR::IntegerConstant_t>(args[1])->m_n;
         int kind = ASRUtils::extract_kind_from_ttype_t(ASR::down_cast<ASR::IntegerConstant
_t>(args[0])->m_type);
-        int bits_size = 8*kind;
+        bool negative_shift = (shift_signed < 0);
+        uint32_t shift = abs(shift_signed);
+        uint32_t bits_size = 8u * (uint32_t)kind;
+        uint32_t max_bits_size = 64;
         if (bits_size < shift) {
             throw SemanticError("The absolute value of SHIFT argument must be less than or
 equal to BIT_SIZE('I')",
             args[0]->base.loc);
         }
-        int64_t result;
-        if (shift < 0) {
-            shift = -shift;  
-            result = (val >> shift) | (val << (bits_size - shift));
+        val = cutoff_extra_bits(val, bits_size, max_bits_size);
+        uint64_t result;
+        if (negative_shift) {
+            result = (val >> shift)
+                    | cutoff_extra_bits(val << (bits_size - shift), bits_size, max_bits_si
ze);
         } else {
-            result = (val << shift) | (val >> (bits_size - shift)); 
+            result = cutoff_extra_bits(val << shift, bits_size, max_bits_size)
+                    | ((val >> (bits_size - shift)));
         }
         return make_ConstantWithType(make_IntegerConstant_t, result, t1, loc);
     }
```
```console
% cat examples/expr2.f90 
program expr2

    print *, ishftc(10_1, 2_1)
    print *, ishftc(10_1, -2_1)
    print *, ishftc(-10_1, 2_1)
    print *, ishftc(-10_1, -2_1)

    print *, ishftc(10_2, 2_2)
    print *, ishftc(10_2, -2_2)
    print *, ishftc(-10_2, 2_2)
    print *, ishftc(-10_2, -2_2)

    print *, ishftc(10_4, 2_4)
    print *, ishftc(10_4, -2_4)
    print *, ishftc(-10_4, 2_4)
    print *, ishftc(-10_4, -2_4)

    print *, ishftc(10_8, 2_8)
    print *, ishftc(10_8, -2_8)
    print *, ishftc(-10_8, 2_8)
    print *, ishftc(-10_8, -2_8)

    print *, ishftc(127_1, 7_1)
    print *, ishftc(127_1, -7_1)
    print *, ishftc(-127_1, 7_1)
    print *, ishftc(-127_1, -7_1)

    print *, ishftc(32767_2, 15_2)
    print *, ishftc(32767_2, -15_2)
    print *, ishftc(-32767_2, 15_2)
    print *, ishftc(-32767_2, -15_2)

    print *, ishftc(2147483647_4, 31_4)
    print *, ishftc(2147483647_4, -31_4)
    print *, ishftc(-2147483647_4, 31_4)
    print *, ishftc(-2147483647_4, -31_4)

    print *, ishftc(9223372036854775807_8, 63_8)
    print *, ishftc(9223372036854775807_8, -63_8)
    print *, ishftc(-9223372036854775807_8, 63_8)
    print *, ishftc(-9223372036854775807_8, -63_8)

    print *, ishftc(127_1, 8_1)
    print *, ishftc(127_1, -8_1)
    print *, ishftc(-127_1, 8_1)
    print *, ishftc(-127_1, -8_1)

    print *, ishftc(32767_2, 16_2)
    print *, ishftc(32767_2, -16_2)
    print *, ishftc(-32767_2, 16_2)
    print *, ishftc(-32767_2, -16_2)

    print *, ishftc(2147483647_4, 32_4)
    print *, ishftc(2147483647_4, -32_4)
    print *, ishftc(-2147483647_4, 32_4)
    print *, ishftc(-2147483647_4, -32_4)

    print *, ishftc(9223372036854775807_8, 64_8)
    print *, ishftc(9223372036854775807_8, -64_8)
    print *, ishftc(-9223372036854775807_8, 64_8)
    print *, ishftc(-9223372036854775807_8, -64_8)

end program
% gfortran examples/expr2.f90 
% ./a.out 
   40
 -126
  -37
  -67
     40
 -32766
    -37
 -16387
          40
 -2147483646
         -37
 -1073741827
                   40
 -9223372036854775806
                  -37
 -4611686018427387907
  -65
   -2
  -64
    3
 -16385
     -2
 -16384
      3
 -1073741825
          -2
 -1073741824
           3
 -4611686018427387905
                   -2
 -4611686018427387904
                    3
  127
  127
 -127
 -127
  32767
  32767
 -32767
 -32767
  2147483647
  2147483647
 -2147483647
 -2147483647
  9223372036854775807
  9223372036854775807
 -9223372036854775807
 -9223372036854775807
% lfortran examples/expr2.f90
40
-126
-37
-67
40
-32766
-37
-16387
40
-2147483646
-37
-1073741827
40
-9223372036854775806
-37
-4611686018427387907
-65
-2
-64
3
-16385
-2
-16384
3
-1073741825
-2
-1073741824
3
-4611686018427387905
-2
-4611686018427387904
3
127
127
-127
-127
32767
32767
-32767
-32767
2147483647
2147483647
-2147483647
-2147483647
9223372036854775807
9223372036854775807
-9223372036854775807
-9223372036854775807
```

Comment (Shaikh-Ubaid) [reply src/libasr/pass/intrinsic_function_registry.h:2329]: PS: Minimum changes that make it work.

```diff
diff --git a/src/libasr/pass/intrinsic_function_registry.h b/src/libasr/pass/intrinsic_function_registry.h
index 13e2a9801..819f643bd 100644
--- a/src/libasr/pass/intrinsic_function_registry.h
+++ b/src/libasr/pass/intrinsic_function_registry.h
@@ -2310,6 +2310,13 @@ namespace Idint {
 
 namespace Ishftc {
 
+    static int64_t cutoff_extra_bits(int64_t num, int bits_size, int max_bits_size) {
+        if (bits_size == max_bits_size) {
+            return num;
+        }
+        return (num & ((1LL << bits_size) - 1LL));
+    }
+
     static ASR::expr_t *eval_Ishftc(Allocator &al, const Location &loc,
             ASR::ttype_t* t1, Vec<ASR::expr_t*> &args) {
         int64_t val = ASR::down_cast<ASR::IntegerConstant_t>(args[0])->m_n;
@@ -2320,12 +2327,13 @@ namespace Ishftc {
             throw SemanticError("The absolute value of SHIFT argument must be less than or equal to BIT_SIZE('I')",
             args[0]->base.loc);
         }
+        val = cutoff_extra_bits(val, bits_size, 64);
         int64_t result;
         if (shift < 0) {
             shift = -shift;
-            result = (val >> shift) | (val << (bits_size - shift));
+            result = ((uint64_t)val >> shift) | cutoff_extra_bits(val << (bits_size - shif
t), bits_size, 64);
         } else {
-            result = (val << shift) | (val >> (bits_size - shift)); 
+            result = cutoff_extra_bits(val << shift, bits_size, 64) | ((uint64_t)val >> (b
its_size - shift));
         }
         return make_ConstantWithType(make_IntegerConstant_t, result, t1, loc);
     }
```

Comment (Pranavchiku) [reply src/libasr/pass/intrinsic_function_registry.h:2329]: Nice. I am wondering about how we will cutoff extra bits at runtime. Looks slightly tricky.

File: src/libasr/pass/intrinsic_function_registry.h [hunk 16]
```diff
@@ -2306,6 +2308,77 @@ namespace Idint {
...
 
+namespace Ishftc {
+
+    static ASR::expr_t *eval_Ishftc(Allocator &al, const Location &loc,
+            ASR::ttype_t* t1, Vec<ASR::expr_t*> &args) {
+        int64_t val = ASR::down_cast<ASR::IntegerConstant_t>(args[0])->m_n;
+        int64_t shift = ASR::down_cast<ASR::IntegerConstant_t>(args[1])->m_n;
+        int kind = ASRUtils::extract_kind_from_ttype_t(ASR::down_cast<ASR::IntegerConstant_t>(args[0])->m_type);
+        int bits_size = 8*kind;
+        if (bits_size < shift) {
```
Comment (Shaikh-Ubaid) [inline src/libasr/pass/intrinsic_function_registry.h:2319]: `shift` here can be negative. So we should take absolute value.
```suggestion
        if (bits_size < abs(shift)) {
```

Also, I think the condition is more readable when it is written as `if (abs(shift) > bits_size)`

File: src/libasr/pass/intrinsic_function_registry.h (hunk 15 above)
Comment (Shaikh-Ubaid) [reply src/libasr/pass/intrinsic_function_registry.h:2329]: I think one of the ways could be that `instantiate_Ishftc()` defines/adds two functions `"_lcompilers_Ishftc_" + type` and `cutoff_extra_bits` (we might need to do name mangling or reuse if this is already defined in the same scope). Another way could be we try to expand the code of the function where it is called (like inlining it). For example: 

`val = cutoff_extra_bits(val, bits_size, 64);` can be written as 

```cpp
if (bits_size == max_bits_size) {
    val = val;
} else {
    val = (val & ((1LL << bits_size) - 1LL));
}
```

For other cases like ` result = cutoff_extra_bits(val << shift, bits_size, 64) | ...`, we might need to create extra variables.

Comment (certik) [reply src/libasr/pass/intrinsic_function_registry.h:2329]: This is in the right direction. I still think the implementation is too complicated, let's figure out the simplest clean implementation of this function at compile time using operations that we support, and then let's implement it using ASR. If we need some new operation, we can consider adding it to ASR.

Comment (certik) [reply src/libasr/pass/intrinsic_function_registry.h:2329]: Let's do two branches. One for positive shift, one for negative shift.

In each of the branches, it should be possible to figure out a one liner that does the job. You might need to cutoff the "size" if it is larger than 64, or something like that. 

Furthermore we need to get this working for all integer sizes. 

But I think you simply do some simple if checks, and then the bit rotation should be a one liner.

Comment (certik) [reply src/libasr/pass/intrinsic_function_registry.h:2329]: Here is the basic idea:
```c
((value) << (shift)) | ((value) >> ((size) - (shift)))
```
You take the value, and do two things:
* shift the main bits (getting rid of the bits that fall off)
* shift the bits that fall off to the other side
And you "or" the two results, which effectively creates a rotation.

Just the details have to be polished and all corner cases covered.

**Comment by certik** [comment]
@parth121101 what's the status of this --- let's finish it?

**Comment by parth121101** [comment]
@certik I was working on some other functions. But surely I would look into this PR and complete it within 1-2 days.

**Comment by certik** [comment]
Perfect!

**Comment by certik** [comment]
To make progress, let's give a compiler error for the runtime implementation / instantiation. And first focus on compile time implementation and let's ensure it passes all the tests. Once this works correctly and agrees with GFortran, we can merge it. Then later we can implement the runtime implementation.

**Comment by parth121101** [comment]
The implementation of this function is continued in the PR: #3523
